### PR TITLE
[BEAM-11105] Add manual watermark estimation

### PR DESCRIPTION
**Pull request by damccorm**
**Summary of Changes**

This pr is a continuation of building watermark support in the Go Sdk. It builds on the work done in https://github.com/apache/beam/pull/17374 and introduces the ability for watermark estimators to be invoked from ProcessElement. It includes the following changes:

- Validation (fn/fn.go, fn/fn_test.go, graph/fn.go, graph/fn_test.go)
- Execution (exec/fn.go and exec/pardo.go)
- Code generation (exec/fn_arity.tmpl, exec/fn_arity.go, reflectx/calls.tmpl, reflectx/calls.go)
- Concrete implementation of a manual estimator (sdf/watermark_estimator.go)

This is built according to the design specified here - https://docs.google.com/document/d/1DqCYJ-J1YGNelCRIcN5v6BQsZxJB2l5uWBmN4ti--Ew/edit?usp=sharing

**Next Steps**

The following work remains to be implemented after this is completed

- Timestamp observing estimators (call ObserveTimestamp on each emitted timestamp) - in progress, https://github.com/apache/beam/pull/17476
- Doc changes

**Manual Testing**

On top of all the unit tests introduced, I hacked the existing textio sdf doFn and added the following function:

```
func (fn *readSdfFn) CreateWatermarkEstimator(t time.Time) *sdf.ManualWatermarkEstimator {
	return &sdf.ManualWatermarkEstimator{State: t}
}

func (fn *readSdfFn) InitialWatermarkEstimatorState(_ beam.EventTime, _ offsetrange.Restriction, _ string, _ int64) time.Time {
	return time.Now()
}

func (fn *readSdfFn) WatermarkEstimatorState(e *sdf.ManualWatermarkEstimator) time.Time {
	return e.State
}
```

and added the following to ProcessElement:

```
func (fn *readSdfFn) ProcessElement(ctx context.Context, we *sdf.ManualWatermarkEstimator, rt *sdf.LockRTracker, filename string, _ int64, emit func(string)) error {
	we.UpdateWatermark(time.Now().AddDate(0, 1, 0)) // Add 1 month to current time
```

and ran wordCount (with the textio sdf version) on dataflow so that I could observe the output watermark was getting correctly updated to the current wall clock time + 1 month:

image

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by damccorm** [comment]
assign to next reviewer

**Comment by damccorm** [comment]
R: @jrmccluskey @riteshghorse for initial review since Daniel is out

**Comment by damccorm** [comment]
R: @lostluck for final approval

File: sdks/go/pkg/beam/core/runtime/exec/fn_arity.tmpl [hunk 1]
```diff
@@ -30,7 +30,7 @@ import (
 func (n *invoker) initCall() {
 	switch fn := n.fn.Fn.(type) {
 {{range $out := upto 5}}
-{{range $in := upto 9}}
+{{range $in := upto 10}}
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/exec/fn_arity.tmpl:33]: Do we still see benefits of increasing this in micro benchmarks? 

eg. Is manually unrolling args as we do for `Call9x?` methods showing benefit vs passing them in as a slice to `Call` ?

**Review by lostluck — COMMENTED** [review]
The watermark code is fine, but I question the need to continue to add additional parameter arities to the reflectx.Func interfaces. 

Please add a benchmark with 9 vs the general Call to validate it's still worthwhile before we commit this to the repo.

File: sdks/go/pkg/beam/core/runtime/exec/fn_arity.tmpl (hunk 1 above)
Comment (damccorm) [reply sdks/go/pkg/beam/core/runtime/exec/fn_arity.tmpl:33]: Yeah, that's a good call - there was not a clear benefit to doing this. I tried the following benchmarks:

```
func BenchmarkInvokeWith9Args_Call9x4(b *testing.B) {
	fn, _ := funcx.New(reflectx.MakeFunc9x4(func(a1 typex.EventTime, a2 bool, a3 string, a4 int, a5 bool, a6 int, a7 string, a8 bool, a9 int) (int, bool, string, error) {
		return a4 + a6 + a9, a2 && a5 && a8, a3 + a7, nil
	}))
	args := []interface{}{mtime.ZeroTimestamp.Add(2 * time.Millisecond), true, "hello", 1, true, 3, "world", false, 4}
	f := fn.Fn.(reflectx.Func9x4)
	for n := 0; n < b.N; n++ {
		f.Call9x4(args[0], args[1], args[2], args[3], args[4], args[5], args[6], args[7], args[8])
	}
}

func BenchmarkInvokeWith9Args_Call(b *testing.B) {
	fn, _ := funcx.New(reflectx.MakeFunc(func(a1 typex.EventTime, a2 bool, a3 string, a4 int, a5 bool, a6 int, a7 string, a8 bool, a9 int) (int, bool, string, error) {
		return a4 + a6 + a9, a2 && a5 && a8, a3 + a7, nil
	}))
	args := []interface{}{mtime.ZeroTimestamp.Add(2 * time.Millisecond), true, "hello", 1, true, 3, "world", false, 4}
	f := fn.Fn
	for n := 0; n < b.N; n++ {
		f.Call(args)
	}
}
```

with the result:

```
BenchmarkInvokeWith9Args_Call9x4-16    	  722283	      1577 ns/op
BenchmarkInvokeWith9Args_Call-16       	  805590	      1447 ns/op
```

So the normal call is actually more performant. It is probably worth revisiting this more systematically for lower arities as well, but I think that should fall outside the scope of this PR. For the moment, I've lowered it back down to 9

Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/exec/fn_arity.tmpl]: SGTM

**Review by lostluck — APPROVED** [review]
LGTM Thanks!

**Pull request merged**